extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java

```diff
@@ -0,0 +1,196 @@
+  {
+    return valueParser.read().map(
+        r -> {
+          MapBasedInputRow valueRow;
+          try {
+            // Return type for the value parser should be of type MapBasedInputRow
+            // Parsers returning other types are not compatible currently.
+            valueRow = (MapBasedInputRow) r;
+          }
+          catch (ClassCastException e) {
+            throw new ParseException("Unsupported input format in valueFormat. KafkaInputformat only supports input format that return MapBasedInputRow rows");
+          }
+          Map<String, Object> event = new HashMap<>(headerKeyList);
+          /* Currently we prefer payload attributes if there is a collision in names.
+              We can change this beahvior in later changes with a config knob. This default
+              behavior lets easy porting of existing inputFormats to the new one without any changes.
+            */
+          event.putAll(valueRow.getEvent());
+
+          HashSet<String> newDimensions = new HashSet<String>(valueRow.getDimensions());
+          newDimensions.addAll(headerKeyList.keySet());
+          // Remove the dummy timestamp added in KafkaInputFormat
+          newDimensions.remove("__kif_auto_timestamp");
+
+          final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+          final List<String> dimensions;
+          if (!schemaDimensions.isEmpty()) {
+            dimensions = schemaDimensions;
+          } else {
+            dimensions = Lists.newArrayList(
+                Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+            );
+          }
+
+          return new MapBasedInputRow(
+              inputRowSchema.getTimestampSpec().extractTimestamp(event),
+              dimensions,
+              event
+          );
+        }
+    );
+  }
+
+  private CloseableIterator<InputRow> buildRowsWithoutValuePayload(Map<String, Object> headerKeyList)
+  {
+    HashSet<String> newDimensions = new HashSet<String>(headerKeyList.keySet());
+    final List<String> schemaDimensions = inputRowSchema.getDimensionsSpec().getDimensionNames();
+    final List<String> dimensions;
+    if (!schemaDimensions.isEmpty()) {
+      dimensions = schemaDimensions;
+    } else {
+      dimensions = Lists.newArrayList(
+          Sets.difference(newDimensions, inputRowSchema.getDimensionsSpec().getDimensionExclusions())
+      );
+    }
+    InputRow row = new MapBasedInputRow(
+        inputRowSchema.getTimestampSpec().extractTimestamp(headerKeyList),
+        dimensions,
+        headerKeyList
+    );
+    List<InputRow> rows = Collections.singletonList(row);
+    return CloseableIterators.withEmptyBaggage(rows.iterator());
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read() throws IOException
+  {
+    Map<String, Object> mergeList = new HashMap<>();
```

Comment:
nit, calling it a list is a bit of a misnomer.

Reply:
done